src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseCaseReport.java

```diff
@@ -0,0 +1,252 @@
+              new ColumnDefinition("18 mo Follow Up Buffy Coat"),
+              new ColumnDefinition("2 yr Follow Up Serum"),
+              new ColumnDefinition("2 yr Follow Up Plasma"),
+              new ColumnDefinition("2 yr Follow Up Buffy Coat"),
+              new ColumnDefinition("2 yr Follow Up Urine Supernatant"),
+              new ColumnDefinition("2 yr Follow Up Urine Pellet"),
+              new ColumnDefinition("Bx Slides - Positive"),
+              new ColumnDefinition("Bx Slides - Negative"),
+              new ColumnDefinition("RP Slides")));
+
+  private enum TimePoint {
+    // Coded by the Times Received portion of the tissue name: PRE_1111111_Xx_X_nn_*#*-1
+    RANDOMIZATION("Screening/Randomization", 1),
+    SIX("6 mo Follow Up", 2),
+    TWELVE("1 yr Follow Up", 3), //
+    EIGHTEEN("18 mo Follow Up", 4),
+    TWENTY_FOUR("2 yr Follow Up", 5),
+    BX("Bx Slides", 6),
+    RP("RP Slides", 7);
+    private final String key;
+    private final int timePointCode;
+    private static final Map<String, PreciseCaseReport.TimePoint> lookup = new TreeMap<>();
+
+    static {
+      for (PreciseCaseReport.TimePoint s : EnumSet.allOf(PreciseCaseReport.TimePoint.class)) {
+        lookup.put(s.getKey(), s);
+      }
+    }
+
+    TimePoint(String key, int timePointCode) {
+      this.key = key;
+      this.timePointCode = timePointCode;
+    }
+
+    public String getKey() {
+      return key;
+    }
+
+    public Predicate<SampleDto> predicate() {
+      return sample -> {
+        if (SAMPLE_CATEGORY_IDENTITY.equals(getAttribute(ATTR_CATEGORY, sample))) return false;
+        return timePointCode == getTimesReceived(sample.getName());
+      };
+    }
+  }
+
+  @Override
+  protected void collectData(PineryClient pinery) throws HttpResponseException, IOException {
+    List<SampleDto> allPreciseSamples =
+        pinery
+            .getSample()
+            .allFiltered(new SampleClient.SamplesFilter().withProjects(Lists.newArrayList("PRE")));
+    Map<String, SampleDto> allPreciseSamplesById = mapSamplesById(allPreciseSamples);
+    Map<String, List<SampleDto>> allPreciseSamplesByIdentity =
+        mapChildrenByIdentityId(allPreciseSamplesById);
+    List<String> tissueOriginOrder = Lists.newArrayList("Se", "Pl", "Ly");
+    List<String> tissueOriginOrderExtra = Lists.newArrayList("Se", "Pl", "Ly", "Us", "Up");
+    allPreciseIdentities =
+        allPreciseSamples
+            .stream()
+            .filter(s -> SAMPLE_CATEGORY_IDENTITY.equals(getAttribute(ATTR_CATEGORY, s)))
+            .collect(Collectors.toList());
+
+    table = new LinkedList<>();
+    List<SampleDto> potentialSamplesAtCurrentTime = null;
+    // One row per identity
+    for (SampleDto identity : allPreciseIdentities) {
+      List<String> row = new LinkedList<>();
+      row.add(identity.getName());
+      potentialSamplesAtCurrentTime = allPreciseSamplesByIdentity.get(identity.getId());
+      if (potentialSamplesAtCurrentTime == null) {
```

Comment:
`potentialSamplesAtCurrentTime` isn't specific to current time, but it's not really used for anything either. Could check `if (allPreciseIdentitiesByIdentity.hasKey(identity.getId))` instead

Reply:
Fixed